### ISSUE TITLE
[bug] Horovod does not work with PyTorch

### ISSUE THREAD

**Issue by bveeramani** [aws/deep-learning-containers#553]
This issue is a follow-up to https://github.com/aws/sagemaker-pytorch-training-toolkit/pull/217. Most of the work for addressing this issue has already been implemented in [this draft PR](https://github.com/aws/deep-learning-containers/pull/346).

Checklist
- [x] I've prepended issue tag with type of change: [bug]
- [x] (If applicable) I've attached the script to reproduce the bug
- [x] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [x] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] I'm using an existing DLC image listed here: https://docs.aws.amazon.com/deep-learning-containers/latest/devguide/deep-learning-containers-images.html
- [x] I've built my own container based off DLC (and I've attached the code used to build my own image)

## Concise Description
Horovod does not work with the PyTorch images.

## DLC image/dockerfile:
The PyTorch GPU training Dockerfiles are most relevant to this issue. 

The CPU Dockerfiles could also be updated. But, as far as I know, people don't usually use Horovod with CPU nodes.

## Current behavior:

I ran into the errors while working with PyTorch 1.4.0 CUDA 10.1 Python 3 image: 

```
763104351884.dkr.ecr.$region.amazonaws.com/pytorch-training:1.4.0-gpu-py3
```

In particular, I was running the tests in the PyTorch training toolkit (which used [this Dockerfile](https://github.com/aws/sagemaker-pytorch-training-toolkit/blob/6afc3ef9524e6c9d0ee30299d9a81d1d3c27a65a/test/container/1.4.0/Dockerfile.dlc.gpu)). I haven't reproduced the errors with any of the other framework versions, but I assume the issues would arise with them as well.

**Error 1: mpi4py is not installed**
Using the PyTorch training image, run the [Horovod integration tests](https://github.com/aws/sagemaker-pytorch-training-toolkit/blob/master/test/integration/local/test_horovod.py) found in the [PyTorch training toolkit repository](https://github.com/aws/sagemaker-pytorch-training-toolkit). You should receive the following error:

```
[1,0]<stderr>:/opt/conda/bin/python: No module named mpi4py
[1,1]<stderr>:/opt/conda/bin/python: No module named mpi4py
```

The error can be resolved by adding the following line to the Dockerfile:

```
RUN pip3 install mpi4py==3.0.3
```

**Error 2: SSH login**
In addition to the preceding error, I have also received the following error:

```
2020-08-27 01:45:04,050 sagemaker-training-toolkit ERROR    framework error: 
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/sagemaker_training/trainer.py", line 85, in train
    entrypoint()
  File "/opt/conda/lib/python3.7/site-packages/sagemaker_pytorch_container/training.py", line 108, in main
    train(environment.Environment())
  File "/opt/conda/lib/python3.7/site-packages/sagemaker_pytorch_container/training.py", line 62, in train
    runner_type=runner_type)
  File "/opt/conda/lib/python3.7/site-packages/sagemaker_training/entry_point.py", line 100, in run
    wait, capture_error
  File "/opt/conda/lib/python3.7/site-packages/sagemaker_training/process.py", line 150, in run
    self._setup()
  File "/opt/conda/lib/python3.7/site-packages/sagemaker_training/mpi.py", line 158, in _setup
    _start_sshd_daemon()
  File "/opt/conda/lib/python3.7/site-packages/sagemaker_training/mpi.py", line 280, in _start_sshd_daemon
    raise RuntimeError(_SSH_DAEMON_NOT_FOUND_ERROR_MESSAGE)
RuntimeError: 
SSH daemon not found, please install SSH to allow MPI to communicate different nodes in cluster.

You can install ssh by running following commands:
-------------------------------------------------

1. Install SSH via apt-get:

apt-get update && apt-get install -y --no-install-recommends openssh-server && mkdir -p /var/run/sshd

2. SSH login fix. Otherwise user is kicked off after login:
sed 's@session\s*required\s*pam_loginuid.so@session optional pam_loginuid.so@g' -i /etc/pam.d/sshd

3. Create SSH key to allow password less ssh between different docker instances:
mkdir -p /root/.ssh/ && ssh-keygen -q -t rsa -N '' -f /root/.ssh/id_rsa &&   cp /root/.ssh/id_rsa.pub /root/.ssh/authorized_keys &&   printf "Host *
  StrictHostKeyChecking no
" >> /root/.ssh/config



SSH daemon not found, please install SSH to allow MPI to communicate different nodes in cluster.

You can install ssh by running following commands:
-------------------------------------------------

1. Install SSH via apt-get:

apt-get update && apt-get install -y --no-install-recommends openssh-server && mkdir -p /var/run/sshd

2. SSH login fix. Otherwise user is kicked off after login:
sed 's@session\s*required\s*pam_loginuid.so@session optional pam_loginuid.so@g' -i /etc/pam.d/sshd

3. Create SSH key to allow password less ssh between different docker instances:
mkdir -p /root/.ssh/ && ssh-keygen -q -t rsa -N '' -f /root/.ssh/id_rsa &&   cp /root/.ssh/id_rsa.pub /root/.ssh/authorized_keys &&   printf "Host *
  StrictHostKeyChecking no
" >> /root/.ssh/config
```

It seems that adding

```
RUN mkdir -p /var/run/sshd && \
 sed 's@session\s*required\s*pam_loginuid.so@session optional pam_loginuid.so@g' -i /etc/pam.d/sshd

RUN rm -rf /root/.ssh/ && \
 mkdir -p /root/.ssh/ && \
 ssh-keygen -q -t rsa -N '' -f /root/.ssh/id_rsa && \
 cp /root/.ssh/id_rsa.pub /root/.ssh/authorized_keys \
 && printf "Host *\n StrictHostKeyChecking no\n" >> /root/.ssh/config
```

to the Dockerfile resolves the error, although I'm not entirely sure if it's necessary for all of the PyTorch images. The preceding code snippet was retrieved from [this MXNet DLC PR](https://github.com/aws/deep-learning-containers/pull/273).

**Error 3: Mutually exclusive MCA settings**
Another issue I have received is that the MCA settings passed in by the [`MPIRunnerType`](https://github.com/aws/sagemaker-training-toolkit/blob/master/src/sagemaker_training/mpi.py#L191-L234) are mutually exclusive with the MCA settings in the config. The resulting error looks something like this:

```
Two mutually-exclusive MCA variables were specified.  This can result
in undefined behavior, such as ignoring the components that the MCA
variables are supposed to affect.

  1st MCA variable: btl_tcp_if_include
    Source of value: environment
  2nd MCA variable: btl_tcp_if_exclude
    Source of value: file (/home/.openmpi/etc/openmpi-mca-params.conf:62)
```

You can get around this error by commenting out the specific line in the MPI config that specifies the MCA setting:

```
RUN sed -i '62,62 s/^/#/' /home/.openmpi/etc/openmpi-mca-params.conf
```

It's pretty hack-ish, but it fixed the issue when I was implemented the [MPIRunnerType PR](https://github.com/aws/deep-learning-containers/pull/346). There's probably a better way to fix the error.

## Expected behavior
The tests should pass without errors.

## Additional context

Here's what still needs to be done before merging [this PR](https://github.com/aws/deep-learning-containers/pull/346):
- Verify that each of the three proposed fixes described above are necessary for all of the PyTorch images.
- If some of the changes are unnecessary, remove them from the appropriate Dockerfiles in the [horovod-pt](https://github.com/harshp8l/deep-learning-containers/tree/horovod-pt) branch.
- Fix any issues with the Horovod tests in the [horovod-pt](https://github.com/harshp8l/deep-learning-containers/tree/horovod-pt) branch. The tests should logically be okay, but there may be some minor syntax-esque errors.
- Remove to-do items from the [Dockerfile.dlc.gpu](https://github.com/aws/sagemaker-pytorch-training-toolkit/blob/master/test/container/1.4.0/Dockerfile.dlc.gpu#L4-L20) file in the PyTorch training kit repository once the PyTorch images are updated.

**Comment by ChaiBapchya** [comment]
Has the sagemaker-python-sdk updated to accept mpi in distributions parameter for PyTorch? 
```
distributions={"mpi":{"enabled":True}}
```
or will it be used using
```
hyperparameters={sagemaker_mpi_enabled:True}
```

@harshp8l @saimidu @bveeramani @ChuyangDeng

**Comment by austinmw** [comment]
@ChaiBapchya Any updates on this issue? 
I got this to work by extending the PyTorch training container with these changes (with Horovod 0.20.3), and then subclassing the PyTorch estimator to create a CustomPyTorch estimator with a distribution parameter and configuration method.

**Comment by ChaiBapchya** [comment]
Right. the whole purpose of this issue was so that: user doesnt have to do A,B,C to get horovod to work with PT.

**Comment by austinmw** [comment]
Is there any progress on this? I'm seeing several customers trying to get pytorch distributed working and running into this blocker.

**Comment by ChaiBapchya** [comment]
by December you should have built-in support for horovod with atleast 1 set of PT images for training.

**Comment by austinmw** [comment]
I wonder if the sagemaker-examples pytorch-horovod-mnist which currently does not function should be removed in the meantime.

**Comment by ChaiBapchya** [comment]
Feel free to create issue with the specific errors on the SM examples repo & the respective framework teams can take a look and fix it.

**Comment by arjkesh** [comment]
Hi @bveeramani - we have added mpi4py and verified horovod use cases on gpu containers after PT 1.6. We are closing this issue for now, but please let us know if you encounter anything else and feel free to reopen. Thank you.
